tests/Loader/LoaderIntegrationTest.php

```diff
@@ -1861,12 +1861,12 @@ public function provideFixturesToGenerate()
             [
                 'parameters' => [],
                 'objects' => [
-                    'another_dummy' => $anotherDummy1 = StdClassFactory::create([
-                        'val' => 1,
-                    ]),
-                    'dummy' => $dummy1 = StdClassFactory::create([
-                        'val' => $anotherDummy1,
-                    ]),
+                    'another_dummy' => $anotherDummy1 = (function (FixtureEntity\DummyWithThrowableSetter $anotherDummy1) {
```

Comment:
_Maybe_ the reason why the tests still pass nonetheless are that I think this test is incomplete. The bug is that in this context, the object (`$anotherDummy`) is getting initialised + hydrated twice (instead of once — this is the bug) + getting the function calls.

I think the current tests don't account for the last call, so what I'm afraid is happening right now is initialised + hydrated (once — fix) ~~+ getting the function class~~ (and this is not good)

So if my assumption is correct, if we add a function call, e.g. `DummyWithThrowableSetter::complete()` to `$anoherDummy1` the tests would fail again with your current patch.

Provided the above is correct, we need to find something else to fix this. I can't take a look at the moment, but the issue is that `$anotherDummy1` is required at some point, it is getting initialised and everything, but for some reason even though it has been created as a complete object it has not been marked as such.

Reply:
sry, i've lost u ... 

i've nothing changed on the function calls? they should still be called? or am i wrong? from the point of the original https://github.com/nelmio/alice/issues/851 issue, i think the problem does not occur because of the use of __call but because of the use of the __construct .

i'm not sure about ur point i think i did not understand it correctly. nonetheless i added another test case, maybe this is going the direction ur talking about?

Reply:
Ok let me rephrase it.

So we have two things here: the objects and the fixtures. The fixture describes what the object looks like (how it's constructed, initialised etc.). So what alice do is:

- go through each fixture
- instantiate the object for each fixture (this is the first pass in the objects generator)
- go through each fixture again an "complete" the objects which involves two steps:
    - hydrating the object
    - configuring the object (doing the calls registered in `__call`)
    - mark the object as complete

So an object is complete only when it went through instantiated + hydrated + configured.

However things are not so simple, because an object may need another. For example here `$dummy` needs `$anotherDummy1`. When this happens, we try to get `$anotherDummy1` as a _complete_ object. At this point of time `$anotherDummy1` may not be created or created but not completed yet. So we need to complete `$anotherDummy1` first before passing it to `$dummy`.

I think the bug here is that when `$dummy` needs `$anotherDummy1`, we do complete it properly and then pass it to `$dummy`. The bug however is that when we do that, we fail to mark `$anotherDummy1` as a complete object (i.e. make it an instance of `CompleteObject` instead of `SimpleObject`). So later during the second pass, `$anotherDummy1` is being completed again which is the bug

Reply:
ok. but what u are writing is what i was trying to say right from the beginning :) after diggin deep into the code, exactly what u say is what i could reproduce... 

and because of this i do make the object an instance of CompleteObject so that it is not completed again in second pass because everthing of this happend in the first pass...

so i think we are both talking same language here? what is it that makes u unsatisfied ?

Reply:
Oh I'm not unsatisfied at all :) I just wanna make sure we are actually fixing this and not causing another issue at the same time.

So what I think is missing in the test is a `__call` on `$anotherDummy1` like you have for hydrating its value. This is just to make sure that `$antoherDummy1` is properly passed as complete (it gets both hydrated & configured) and that no second call is made for it. To rephrase: you added a test accounting for the hydration, I would add a second for the "config" part.